subprojects/problems/src/main/java/org/gradle/api/problems/interfaces/ProblemBuilder.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.problems.interfaces;
+
+import org.gradle.api.Incubating;
+
+import javax.annotation.Nullable;
+
+/**
+ * A builder interface for {@link Problem} instances.
+ *
+ * @since 8.3
+ */
+@Incubating
+public interface ProblemBuilder {
```

Comment:
🤔 It might be nice to use a Stepwise builder here, so that calls to the API guide you in adding each set of alternatives (location/noLocation, documentedAt/undocumented).  It Could also make usages of this API appear very standardized, as the method calls to this build would all follow the same order.